Unify end/end2/end3/return

mrgriffin opened the pull request:
`end2` is for scripts started with `BattleScriptExecute` and `end3` is for scripts started with `BattleScriptPushCursorAndCallback`. So we make those functions register the behavior they want when a plain `end` is encountered.

`return` on an empty stack behaves like `end`, as it does in the overworld scripts.

Also introduced a `BattleScriptCall(...);` function to replace `BattleScriptPushCursor(); gBattlescriptCurrInstr = ...;`.

mrgriffin commented:
Sadly there are some instances of battle scripts that don't obviously unify, e.g. `BattleScript_BerryStatRaiseEnd` references `BS_ATTACKER` whereas `BattleScript_BerryStatRaiseRet` references `BS_SCRIPTING`.

But there's still a decent number of scripts that can be unified :)

mrgriffin commented:
Ready for review.

These four things cover the changed commands:
- `end2` ⇒ `end`: `BattleScript_TailwindEnds` (via "Tailwind applies for 4 turns").
- `end2` ⇒ `return`: `BattleScript_BerryCurPar` (via `test/hold_effect_cur_par.c`).
- `end3` ⇒ `end`: `BattleScript_PastelVeilActivates` (via `test/ability_pastel_veil.c`).
- `end3` ⇒ `return`: `BattleScript_BlockedByPrimalWeather` (via "Drizzle is prevented by primal weather").

DizzyEggg commented on src/battle_util.c:
```diff
@@ -6418,7 +6375,10 @@ void BattleScriptPushCursorAndCallback(const u8 *BS_ptr)
 {
     BattleScriptPushCursor();
```
Couldn't BattleScriptCall be used here as well?

mrgriffin replied:
Definitely can! Done :)

DizzyEggg reviewed (changes requested):
All in all, this PR looks great. This was a wonderful idea to unify these commands and create a much needed BattleScriptCall.
Code-wise it looks fine, but I'll run some tests in-game to make sure nothing breaks.

DizzyEggg reviewed (changes requested):
I came across a wild bug here. It doesn't happen on upcoming, so it must've been caused by changes in this PR.

1. My first mon is Kyorge with Drizzle
2. Opponent uses Roar
3. Things break(see gif)
![drizzle_roar](https://user-images.githubusercontent.com/16259973/219946048-c69511bb-3666-43ed-9e19-ac5eca9b639c.gif)

mrgriffin commented:
> I came across a wild bug here. It doesn't happen on upcoming, so it must've been caused by changes in this PR.
Fixed! Caused by `BattleScriptPushCursorAndCallback` recording the wrong number for the stack size (an off-by-one). I've added your example in `test/2701.c` :)

AsparagusEduardo commented:
I think I would like to release 1.4.0 before this PR is merged, just in case more stuff is broken

DizzyEggg reviewed (changes requested):
Switch-in abilities are not working. Here's an example with Minior's Shields Down.
![Minior](https://user-images.githubusercontent.com/16259973/220289479-9c3b2184-ac5d-4e32-ac92-53f0098cb8b8.gif)

mrgriffin commented:
> Switch-in abilities are not working. Here's an example with Minior's Shields Down.

Good find! This one is interesting:
- At the start of the first battle the instruction pointer is `NULL`.
- Switch-in abilities use `BattleScriptCall`.
- I tried to clean up switch-in abilities to use the `return` script instead of the `end` script.
Together, this means that we try to `return` to a `NULL` instruction pointer. So what I've done is initialize `gBattlescriptCurrInstr` to point at an `end` at the start of a battle.

_However_, I think there's a decent argument that this PR should not actually change any `end`s to `return`? This would mean there's some situations where scripts could be unified, but we could do that unification later, when there's more tests to catch any more examples like this? lmk what you think, it should be easy enough to revert just these changes.

DizzyEggg commented:
> _However_, I think there's a decent argument that this PR should not actually change any `end`s to `return`? This would mean there's some situations where scripts could be unified, but we could do that unification later, when there's more tests to catch any more examples like this? lmk what you think, it should be easy enough to revert just those changes.

I agree with you here. Let's for now just unify the end commands into one and then do the end->return change once we have more battle tests.

mrgriffin commented:
> I agree with you here. Let's for now just unify the end commands into one and then do the end->return change once we have more battle tests.

Ready for re-review. I left the `return` behavior in-place so that it's available for use, but no existing scripts should make use of it.

DizzyEggg commented:
Seems to be working now. 
If anyone else could take a look and play a bit to see if nothing broke, I'm gonna merge this.

DizzyEggg reviewed (changes requested):
Fix conflicts

Bassoonian commented:
Superseded by #4369
